### PR TITLE
Feature/past invites to same events

### PR DESCRIPTION
**Pull request by Ap1234567**
Depends on https://github.com/PythonFreeCourse/calendar/pull/122

### CODE REVIEW

File: tests/test_event.py [hunk 1]
```diff
@@ -17,5 +46,29 @@ def test_eventview_without_id(self, client):
...
+                                          data=different_invitees_event)
+        assert response.status_code == HTTP_303_SEE_OTHER
+        for invitee in CORRECT_EVENT_FORM_DATA["invited"].split(","):
+            assert invitee in response.headers['location']
+
+    def test_eventedit_post_correct(self, event_test_client, user):
+        response = event_test_client.post("/event/edit",
+                                          data=CORRECT_EVENT_FORM_DATA)
+        assert response.status_code == HTTP_303_SEE_OTHER
+        assert '/event/view/' in response.headers['location']
```
Comment (yammesicka) [inline tests/test_event.py:66]: Maybe `url_for`?

File: app/routers/event.py [hunk 2]
```diff
@@ -21,15 +33,64 @@ async def eventedit(request: Request):
...
 
-def create_event(db, title, start, end, owner_id, content=None, location=None):
+def create_event(db, title, start, end, owner_id, content=None,
+                 location=None, invitees=None):
     """Creates an event and an association."""
 
+    if invitees:
+        invitees_str = ','.join(invitees)
+    else:
+        invitees_str = ''
```
Comment (yammesicka) [inline app/routers/event.py:92]: Define this one as the default, and try not to use Hungarian Notation:

```
invitees_concatenated = ''
if invitees:
    invitees_concatenated = ', '.join(invitees)

File: app/internal/event.py [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import re
```
Comment (yammesicka) [inline app/internal/event.py:1]: Probably in the wrong commit :)

File: app/routers/event.py [hunk 4]
```diff
@@ -1,12 +1,24 @@
...
+from app.database.database import get_db
 from app.database.models import Event
+from app.database.models import User
 from app.database.models import UserEvent
 from app.dependencies import templates
+from app.internal.event import validate_zoom_link
 from app.internal.utils import create_model
+from app.routers.user import create_user
+
+VALID_MAIL_REGEX = re.compile(r'^\S+@\S+\.\S+$')
```
Comment (yammesicka) [inline app/routers/event.py:20]: You can use EmailStr of Pydantic (I think we even have a validation method in another file on this project)

File: app/routers/event.py [hunk 5]
```diff
@@ -21,15 +33,64 @@ async def eventedit(request: Request):
                                       {"request": request})
 
 
+@router.post("/edit")
+async def create_new_event(request: Request, session=Depends(get_db)):
```
Comment (yammesicka) [inline app/routers/event.py:37]: This function is a bit too long - can you think of a way to split it?

**Review by yammesicka — CHANGES_REQUESTED** [review]

File: app/routers/event.py (hunk 5 above)
Comment (yammesicka) [reply app/routers/event.py:35]: Please don't close review comments that you haven't addressed yet.
If you don't agree with the comment - please explain why, we can talk about it :)

Marking the comment as resolved makes me go through the whole code again.

**Review by yammesicka — CHANGES_REQUESTED** [review]

File: app/internal/event.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+def get_invited_emails(invited_from_form):
+    invited_emails = []
+    for invited_email in invited_from_form.split(','):
+        invited_email = invited_email.strip()
+        try:
+            validate_email(invited_email, check_deliverability=False)
+            invited_emails.append(invited_email)
+        except EmailSyntaxError:
+            pass
```
Comment (yammesicka) [inline app/internal/event.py:27]: Write to log?

File: app/internal/event.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+    return invited_emails
+
+
+def get_uninvited_regular_emails(session, owner_id, title, invited_emails):
+    regular_invitees = set()
+    for record in session.query(Event).\
```
Comment (yammesicka) [inline app/internal/event.py:34]: Prefer putting it in a variable for readability. Also, prefer splitting lines using parenthesis.

File: app/routers/event.py [hunk 8]
```diff
@@ -1,12 +1,22 @@
...
+from fastapi import Depends
+from fastapi.responses import RedirectResponse
+from starlette.status import HTTP_303_SEE_OTHER
 
+from app.database.database import get_db
 from app.database.models import Event
+from app.database.models import User
 from app.database.models import UserEvent
 from app.dependencies import templates
+from app.internal.event import validate_zoom_link, get_invited_emails, \
```
Comment (yammesicka) [inline app/routers/event.py:15]: Prefer splitting lines using parenthesis

**Review by yammesicka — CHANGES_REQUESTED** [review]

File: app/internal/event.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+def get_invited_emails(invited_from_form):
+    invited_emails = []
+    for invited_email in invited_from_form.split(','):
+        invited_email = invited_email.strip()
+        try:
+            validate_email(invited_email, check_deliverability=False)
+            invited_emails.append(invited_email)
+        except EmailSyntaxError:
+            logging.debug(f'{invited_email} is not a valid email address')
```
Comment (ivarshav) [inline app/internal/event.py:28]: Isn't it an error log?

File: app/routers/event.py [hunk 10]
```diff
@@ -21,15 +31,54 @@ async def eventedit(request: Request):
                                       {"request": request})
 
 
+@router.post("/edit")
+async def create_new_event(request: Request, session=Depends(get_db)):
+    data = await request.form()
+    title = data['title']
+    content = data['description']
+    start = dt.strptime(data['start_date'] + ' ' + data['start_time'],
+                        '%Y-%m-%d %H:%M')
```
Comment (ivarshav) [inline app/routers/event.py:40]: Consider keep the format in a const

File: app/templates/event/eventview.html [hunk 11]
```diff
@@ -29,6 +32,7 @@
                     </div>
                     <!--    Copy commented section to add another tab-->
                     <!--    <div class="tab-pane fade" id="(CHANGE_ME)" role="tabpanel" aria-labelledby="(CHANGE_ME)-tab">-->
+                    <!--    <div class="tab-pane fade" id="(CHANGE_ME)" role="tabpanel" aria-labelledby="(CHANGE_ME)-tab">-->
```
Comment (ivarshav) [inline app/templates/event/eventview.html:35]: Duplicate by mistake?

File: app/internal/event.py [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    return invited_emails
+
+
+def get_uninvited_regular_emails(session, owner_id, title, invited_emails):
+    regular_invitees = set()
+    invitees_query = session.query(Event).with_entities(Event.invitees)
+    similar_events_invitees = invitees_query.filter(Event.owner_id == owner_id,
+                                                    Event.title == title).all()
+    for record in similar_events_invitees:
+        for email in record[0].split(','):
```
Comment (yammesicka) [inline app/internal/event.py:39]: Prefer `.update`

File: app/internal/event.py [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+            regular_invitees.add(email)
+
+    return regular_invitees.difference(set(invited_emails))
```
Comment (yammesicka) [inline app/internal/event.py:42]: Nice trick: you can use `regular_invitees - set(invited_emails)`

File: app/routers/event.py [hunk 14]
```diff
@@ -21,15 +32,54 @@ async def eventedit(request: Request):
...
+
+    if is_zoom:
+        validate_zoom_link(location)
+
+    event = create_event(session, title, start, end, owner_id, content,
+                         location, invited_emails)
+    message = ''
+    if uninvited_contacts:
+        message = f'Forgot to invite {", ".join(uninvited_contacts)} maybe?'
+    return RedirectResponse(f'/event/view/{event.id}?message={message}',
```
Comment (yammesicka) [inline app/routers/event.py:63]: Prefer `url_for`

File: app/internal/event.py [hunk 15]
```diff
@@ -0,0 +1,42 @@
+import logging
+import re
+
+from email_validator import validate_email, EmailSyntaxError
```
Comment (yammesicka) [inline app/internal/event.py:4]: You can use `pydantic` instead :)

File: app/routers/event.py [hunk 16]
```diff
@@ -21,15 +32,54 @@ async def eventedit(request: Request):
...
+    location_type = data['location_type']
+    is_zoom = location_type == 'vc_url'
+    location = data['location']
+
+    invited_emails = get_invited_emails(data['invited'])
+    uninvited_contacts = get_uninvited_regular_emails(session, owner_id,
+                                                      title, invited_emails)
+
+    if is_zoom:
+        validate_zoom_link(location)
```
Comment (yammesicka) [inline app/routers/event.py:56]: Prefer: `raise_if_zoom_link_invalid`

(`Validate` usually returns `True`/`False`)

File: app/routers/event.py [hunk 17]
```diff
@@ -21,15 +32,54 @@ async def eventedit(request: Request):
...
+                                       "message": message})
 
 
-def create_event(db, title, start, end, owner_id, content=None, location=None):
+def create_event(db, title, start, end, owner_id, content=None,
+                 location=None, invitees=None):
     """Creates an event and an association."""
 
+    invitees_concatenated = ''
+    if invitees:
```
Comment (yammesicka) [inline app/routers/event.py:80]: You can remove the if:

```
In [1]: ','.join([])
Out[1]: ''
```

File: app/routers/event.py [hunk 18]
```diff
@@ -21,15 +32,54 @@ async def eventedit(request: Request):
...
+    if is_zoom:
+        validate_zoom_link(location)
+
+    event = create_event(session, title, start, end, owner_id, content,
+                         location, invited_emails)
+    message = ''
+    if uninvited_contacts:
+        message = f'Forgot to invite {", ".join(uninvited_contacts)} maybe?'
+    return RedirectResponse(f'/event/view/{event.id}?message={message}',
+                            status_code=HTTP_303_SEE_OTHER)
```
Comment (yammesicka) [inline app/routers/event.py:64]: Prefer 302

**Review by yammesicka — CHANGES_REQUESTED** [review]

File: app/internal/event.py (hunk 15 above)
Comment (Ap1234567) [reply app/internal/event.py:4]: Could, but why? Pydantic uses this library.

**Comment by yammesicka** [comment]
Please fix conflicts :)

File: app/templates/event/partials/edit_event_details_tab.html [hunk 19]
```diff
@@ -21,6 +21,10 @@
     <input type="text" name="location" placeholder="VC URL/Location">
 </div>
 
+<div class="form_row">
+    <input type="text" name="invited" placeholder="Invited emails, separated by commas">
```
Comment (yammesicka) [inline app/templates/event/partials/edit_event_details_tab.html:25]: Please add a `<label>` for accessibility

File: app/routers/event.py [hunk 20]
```diff
@@ -183,6 +197,7 @@ def create_event(db, title, start, end, owner_id, content=None, location=None):
         content=content,
         owner_id=owner_id,
         location=location,
+        invitees=invitees_concatenated
```
Comment (yammesicka) [inline app/routers/event.py:200]: Add a trailing ,

**Review by yammesicka — CHANGES_REQUESTED** [review]

File: app/templates/event/partials/edit_event_details_tab.html [hunk 21]
```diff
@@ -21,6 +21,11 @@
     <input type="text" name="location" placeholder="VC URL/Location">
 </div>
 
+<div class="form_row">
+    <label>Invited emails: </label>
```
Comment (yammesicka) [inline app/templates/event/partials/edit_event_details_tab.html:25]: You should use `for` to bind the label to the input :)

**Review by yammesicka — CHANGES_REQUESTED** [review]

**Review by yammesicka — APPROVED** [review]

**Pull request merged**